### PR TITLE
WebChat Sample App

### PR DESCRIPTION
**Pull request by WashingtonKayaker**
# New sample app demonstrating how to enable a bot with Web Chat capabilities

This incorporates the new Direct Line functionality that eliminates the need to embed the direct line secret into the html file, but instead includes a middleware method to get a Direct Line token from the secret contained in the appsettings.json file.

### CODE REVIEW

File: samples/javascript_nodejs/49.WebChat-bot/index.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+// Import required bot configuration.
+const ENV_FILE = path.join(__dirname, '.env');
+dotenv.config({ path: ENV_FILE });
+
+// Create HTTP server
+const server = restify.createServer();
+server.listen(process.env.port || process.env.PORT || 3978, () => {
+    console.log(`\n${ server.name } listening to ${ server.url }`);
+    console.log(`\nGet Bot Framework Emulator: https://aka.ms/botframework-emulator`);
+    console.log(`\nTo talk to your bot, open the emulator select "Open Bot"`);
```
[the remark this answers is not in the document]
Comment (WashingtonKayaker) [reply samples/javascript_nodejs/49.WebChat-bot/index.js]: added

File: samples/javascript_nodejs/49.WebChat-bot/index.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+// Listen for incoming requests.
+server.post('/api/messages', (req, res) => {
+    adapter.processActivity(req, res, async (context) => {
+        // Route to main dialog.
+        await bot.run(context);
+    });
+});
+ // This code performs a token request from direct line using your Web Chat secret to generate a DirectLine Token.   
+    // Your client code must provide either a secret or a token to talk to your bot.
```
[the remark this answers is not in the document]
Comment (WashingtonKayaker) [reply samples/javascript_nodejs/49.WebChat-bot/index.js]: fixed

File: samples/javascript_nodejs/49.WebChat-bot/README.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
+# WebChatBot
+
+Bot Framework v4 echo bot sample with Web Chat enabled.
+
+This bot has been created using [Bot Framework](https://dev.botframework.com), it shows how to enable Web Chat in a simple bot that accepts input from the user and echoes it back.
+
+## Prerequisites
+
+- A subscription to [Microsoft Azure](https://portal.azure.com/)
```
[the remark this answers is not in the document]
Comment (WashingtonKayaker) [reply samples/javascript_nodejs/49.WebChat-bot/README.md]: Made several changes, let me know if it looks good or if I missed anything

File: samples/javascript_nodejs/49.WebChat-bot/public/index.html [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+<head>
+<title>Web Chat: Full-featured bundle</title>
+<meta name="viewport" content="width=device-width, initial-scale=1.0">
+<!--
+    This CDN points to the latest official release of Web Chat. If you need to test against Web Chat's latest bits, please refer to pointing to Web Chat's MyGet feed:
+    https://github.com/microsoft/BotFramework-WebChat#how-to-test-with-web-chats-latest-bits
+    -->
+<script src="https://cdn.botframework.com/botframework-webchat/latest/webchat.js"></script>
+<style>
+    html,
```
[the remark this answers is not in the document]
Comment (WashingtonKayaker) [reply samples/javascript_nodejs/49.WebChat-bot/public/index.html]: I agree... I went with how VS Code auto-formats (Ctrl+k Ctrl+f), but I will change it back....

File: samples/javascript_nodejs/49.WebChat-bot/public/index.html [hunk 5]
```diff
@@ -0,0 +1,40 @@
+<!DOCTYPE html>
+<html lang="en-US">
+
+<head>
+<title>Web Chat: Full-featured bundle</title>
```
Comment (compulim) [inline samples/javascript_nodejs/49.WebChat-bot/public/index.html]: Indentation.

HTML use 2-space indentation nowaday.

File: samples/javascript_nodejs/49.WebChat-bot/public/index.html [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    height: 100%;
+    width: 100%;
+    }
+</style>
+</head>
+
+<body>
+<div id="webchat" role="main"></div>
+<script>
+    (async function () {
```
Comment (compulim) [inline samples/javascript_nodejs/49.WebChat-bot/public/index.html]: Use 2-space for indentation.

Comment (compulim) [inline samples/javascript_nodejs/49.WebChat-bot/public/index.html]: Add `'use strict';`.

File: samples/javascript_nodejs/49.WebChat-bot/public/index.html [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    window.WebChat.renderWebChat({
+        directLine: window.WebChat.createDirectLine({ token })
+    }, document.getElementById('webchat'));
+
+    document.querySelector('#webchat > *').focus();
+    })().catch(err => console.error(err));
+</script>
+</body>
+
+</html>
```
Comment (compulim) [inline samples/javascript_nodejs/49.WebChat-bot/public/index.html]: Empty new line before EOF.

File: samples/javascript_nodejs/49.WebChat-bot/public/index.html [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    This CDN points to the latest official release of Web Chat. If you need to test against Web Chat's latest bits, please refer to pointing to Web Chat's MyGet feed:
+    https://github.com/microsoft/BotFramework-WebChat#how-to-test-with-web-chats-latest-bits
+    -->
+<script src="https://cdn.botframework.com/botframework-webchat/latest/webchat.js"></script>
+<style>
+    html, body { height: 100% }
+    body { margin: 0  }
+
+    #webchat {
+    height: 100%;
```
Comment (compulim) [inline samples/javascript_nodejs/49.WebChat-bot/public/index.html]: CSS indentation.

File: samples/javascript_nodejs/49.WebChat-bot/public/index.html (hunk 5 above)
Comment (WashingtonKayaker) [reply samples/javascript_nodejs/49.WebChat-bot/public/index.html]: Fixed.

BTW: Someone needs to tell the VS Code team.  The formatting was done using Ctrl+k,Ctrl+f.

File: samples/javascript_nodejs/49.WebChat-bot/public/index.html (hunk 6 above)
Comment (WashingtonKayaker) [reply samples/javascript_nodejs/49.WebChat-bot/public/index.html]: fixed

File: samples/javascript_nodejs/49.WebChat-bot/public/index.html (hunk 5 above)
Comment (compulim) [reply samples/javascript_nodejs/49.WebChat-bot/public/index.html]: Then, please format it manually by hand.

Comment (WashingtonKayaker) [reply samples/javascript_nodejs/49.WebChat-bot/public/index.html]: Already did, will be checking in soon.

File: samples/javascript_nodejs/49.WebChat-bot/public/index.html (hunk 6 above)
Comment (WashingtonKayaker) [reply samples/javascript_nodejs/49.WebChat-bot/public/index.html]: Added

File: samples/javascript_nodejs/49.WebChat-bot/public/index.html (hunk 7 above)
Comment (WashingtonKayaker) [reply samples/javascript_nodejs/49.WebChat-bot/public/index.html]: added blank line at EOF.

File: samples/javascript_nodejs/49.WebChat-bot/public/index.html (hunk 8 above)
Comment (WashingtonKayaker) [reply samples/javascript_nodejs/49.WebChat-bot/public/index.html]: I indent the 2 lines between the brackets (height: / width:) 2 spaces.

**Comment by compulim** [comment]
What is the purpose of this sample? I read your description but it is not consistent with the source code. I am confused.

If it is talking about Web Chat, please take away the bot code. Web Chat is UI, not a bot. Also, please take away token server.

If it is talking about token, please take away the bot code and the Web Chat. Token server has nothing to deal with both Web Chat and bot code.

Please do not use IIS to host Web Chat. Web Chat is platform-independent. Use as minimal code as possible to host Web Chat, so the user are not locked into a single platform.

One sample is to teach the developers how to do one very specific thing at a time. If you think it is an integration between multiple systems, call it a demo and host it.

File: samples/csharp_dotnetcore/49.WebChat/appsettings.json [hunk 9]
```diff
@@ -0,0 +1,5 @@
+{
+  "MicrosoftAppId": "ace8108b-4332-4e17-b754-50e7bc92fec8",
+  "MicrosoftAppPassword": "EchoWebChatBotPWD",
+  "DirectLineSecret": "aB-IVTTvBIc.g_v3NlsF3R6D8lhrMFIuWrDxgs4P6V4SdjCyStROy8k"
```
Comment (EricDahlvang) [inline samples/csharp_dotnetcore/49.WebChat/appsettings.json]: These shouldn't be filled with actual values, but rather placeholders.

**Comment by johnataylor** [comment]
We don't think we want to take this into our samples set at this stage. We already feel we have quite a lot of samples. If you feel that you are illustrating some important areas that are not covered, please open an issue for the specific cases. Thanks.

**Pull request closed without merging**